### ISSUE TITLE
Reincarnate flatten() as static method

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#241]
The following allows to `flatten()` Monads using a fluent API while being type safe!

```java
interface Monad<M extends Kind<M, ?>, T> extends Kind<M, T> {

    <U> Monad<M, U> flatMap(Function<? super T, ? extends Kind<M, U>> mapper);

    default <U> Monad<M, U> transform(Function<? super Kind<M, T>, ? extends Kind<M, U>> f) {
        return (Monad<M, U>) f.apply(this);
    }

    static <T, M extends Monad<M, ?>> Monad<M, T> flatten(Monad<M, ? extends Monad<M, T>> monad) {
        return monad.flatMap(Function.identity());
    }
}

interface List<T> extends Kind<List<?>, T>, Seq<List<?>, T>, Monad<List<?>, T> {

    @Override
    <U> List<U> flatMap(Function<? super T, ? extends Kind<List<?>, U>> mapper);

    default <U> List<U> transform(Function<? super Kind<List<?>, T>, ? extends Kind<List<?>, U>> f) {
        return (List<U>) f.apply(this);
    }

    static <T> List<T> flatten(List<? extends List<T>> monad) {
        return monad.flatMap(Function.identity());
    }
}
```

then the following is possible:

```java
final List<Integer> list = List.of(1, 2, 3);

// does not compile, which is correct!
list.transform(List::flatten);


final List<List<Integer>> list2 = List.of(List.of(1, 2, 3));

// yay!
final List<Integer> transform = list2.transform(List::flatten);

// yay!
final Monad<Integer, List<?>> flatten = Monad.flatten(list2);

// does not compile, which is correct!
list2.transform(Monad::flatten);
```

w00t!

**Comment by danieldietrich** [comment]
Btw - the examples above are using #242 already.

**Comment by danieldietrich** [comment]
The above code sketches the general solution of moving runtime type information to the static `flatten()` method, which are passed to the instance method `transform()`.

For a correct implementation of `flatten()` we need to differentiate it by enclosing type. It is not sufficient to call `monad.flatMap(Function.identity())`.

**1)** For univalent/one-valued type (see `Valences.Univalent`), like `Option` and `Try`, we want to allow something like

```java
Option<Try<String>> option = ...;
Option<String> result = option.transform(Option.flatten());
```

We can achieve this like so:

```java
static <T> Option<T> flatten(Option<? extends Monad<?, T> & Univalent<T>> option);
```

Then, because Java has no union-types, we need to define:

```java
// Union type of {@code Univalent} and {@code Monad}.
interface UnivalentMonad<M extends UnivalentMonad<M, ?>, T> extends Univalent<T>, Monad<M, T>;

// and then...
static <T> Option<T> flatten(Option<? extends UnivalentMonad<?, T>> option);
```

**2)** For `Traversable` types the following makes sense:

```java
List<Option<String>> list = List.of(new Some<>(1), None.instance());
List<String> result = option.transform(Option.flatten()); // = List(1)
```

We can achieve this like so:

```java
static <T> List<T> flatten(List<? extends Monad<?, T> > list);
```

Just some thoughts...

**Comment by danieldietrich** [comment]
No need for simulated intersection types. The above example works as follows:

```java
static <T extends Monad<?, T> & Univalent<T>> Option<T> flatten(Option<T> option) {
    ...
}
```

**Comment by danieldietrich** [comment]
Given a `Monad<Monad<T>>` we are able to `flatten()` it by calling `flatMap(m -> m)` or `flatMap(Function.identity())`.

It is common behavior of a collection to flatten Iterable elements, i.e.

```java
List.of(Option.of(1), Option.none(), Option.of(2)).flatten() // = List(1, 2)
```

Given the type `List<T>`, we have (simplified here)

```java
<U> List<U> flatMap(Function<? super T, ? extends List<T>>)
```

But in order to use `flatMap()` to `flatten()` a List of `Option`s (or `Try`s, `Either`s, ...) we need

```java
<U> List<U> flatMap(Function<? super T, ? extends Iterable<T>>)
```

(Scala [does this](http://stackoverflow.com/questions/27750046/is-a-collection-with-flatmap-a-monad) also.)

This will break the `Monad` interface of all `Traversable` implementations.

## Any ideas?

Options:

* introduce a `flatMap'(Function<? super T, ? extends Iterable<T>>)`
* remove extension of `Monad`

**Comment by danieldietrich** [comment]
Depends on #254

**Comment by danieldietrich** [comment]
`Monad` removed in favor of `TraversableOnce`

**Comment by danieldietrich** [comment]
The situation changed, the Monad was moved to the javaslang-pure module. I currently see no reason why we should not provide `flatten()` as static method. The contained type should be `? extends Iterable<? extends T>` (or `? extends Iterable<T>`?).

Example:

```java
interface List<T> extends LinearSeq<T> {
    static <T> List<T> flatten(List<? extends Iterable<? extends T>> list) {
        ...
    }
}
```

**Comment by danieldietrich** [comment]
Too complicated to maintain because not part of the type-hierarchy. We will use `flatMap(t -> t)` instead.

**Comment by gerardbosch** [comment]
Hi @danieldietrich The more I learn Vavr the more I love it, thanks!

I think this was not finally implemented. Will something like this work to flatten lists?

![image](https://user-images.githubusercontent.com/30733556/63603313-b65a9980-c5c9-11e9-9469-ae71a45c97e8.png)

Maybe `List` could be `Seq` ? I can't get to see why this can or cannot work. Thx a lot!
